synapse/rest/client/read_marker.py

```diff
@@ -48,34 +49,51 @@ async def on_POST(
         await self.presence_handler.bump_presence_active_time(requester.user)
 
         body = parse_json_object_from_request(request)
-        read_event_id = body.get(ReceiptTypes.READ, None)
-        hidden = body.get(ReadReceiptEventFields.MSC2285_HIDDEN, False)
-
-        if not isinstance(hidden, bool):
-            raise SynapseError(
-                400,
-                "Param %s must be a boolean, if given"
-                % ReadReceiptEventFields.MSC2285_HIDDEN,
-                Codes.BAD_JSON,
-            )
 
+        read_event_id = body.get(ReceiptTypes.READ, None)
         if read_event_id:
             await self.receipts_handler.received_client_receipt(
                 room_id,
                 ReceiptTypes.READ,
                 user_id=requester.user.to_string(),
                 event_id=read_event_id,
-                hidden=hidden,
             )
 
-        read_marker_event_id = body.get("m.fully_read", None)
+        read_private_event_id = body.get(ReceiptTypes.READ_PRIVATE, None)
+        if read_private_event_id and self.config.experimental.msc2285_enabled:
+            await self.receipts_handler.received_client_receipt(
+                room_id,
+                ReceiptTypes.READ_PRIVATE,
+                user_id=requester.user.to_string(),
+                event_id=read_private_event_id,
+            )
+
+        read_marker_event_id = body.get(ReceiptTypes.FULLY_READ, None)
         if read_marker_event_id:
             await self.read_marker_handler.received_client_read_marker(
                 room_id,
                 user_id=requester.user.to_string(),
                 event_id=read_marker_event_id,
             )
 
+        for key in body.keys():
```

Comment:
This is somewhat optional, but I wonder if it would be clearer in here to:

1. Calculate `valid_receipt_types = {READ}`, if `msc2285_enabled` then add `READ_PRIVATE`.
2. Iterate over the validate receipt types fetching from the body and calling `received_client_read_marker`.
3. Doing the assertion that `set(body.keys()) <= (valid_receipt_types | {FULLY_READ})`

I think we also want the assertion to happen before we persist anything!

Reply:
Thank you for the suggestion - it makes the code much clearer. I ended up not doing number 2 since we sometimes call `received_client_read_marker()` and sometimes `received_client_receipt()`, so it seems iterating would look a little weird; it might make more sense if there is more receipt types added